README
FUSE-Db connector design document
==================================

This is the design document for MySQL connector to the filesystem
written using the FUSE (Filesystem-in-UserSpace-Environment)
library. MySQL FUSE connector is the standard Linux application
that's has been both tested and developed on Fedora-13 i386 box
by Michal Novotny in order to provide facility to access the
data in MySQL database like they were standard directories or
files.

The connector is creating 4 levels on the mountpoint where
3 levels are directory levels (levels 1 to 3) and one level
is the file level (level 4). Both databases and tables are
presented using directory entries as well as primary key
value is being presented on 3rd level. If there's no primary
key in the table this 3rd level directory is not having the
executable bit set which means entering such a directory is
disabled.

On the 4th level the database table columns are being shown
as files. On a new file creation a new field named the same
as the file name is being created in the database. Be aware
of the fact that deleting the files just sets the particular
column in the table to NULL value *not deleting it at all*.

There's a short summary of functionality.

Levels and their meaning are as follows:
 .../DBNames/Table/PKValue/column  (path)
  0      1     2      3      4     (level)

Read implementation for supported levels with corresponding queries:
 - level 1 -> SHOW DATABASES
 - level 2 -> SHOW TABLES
 - level 3 -> SELECT ${PRIMARY_KEY} FROM ${TABLE}
 - level 4 -> SHOW FIELDS FROM ${TABLE}

Write implementation for supported levels with corresponding queries:
 - level 1 -> CREATE DATABASE
 - level 2 -> CREATE TABLE WITH VARCHAR(255) PRIMARY KEY
 - level 3 -> INSERT INTO ${TABLE}(${PK}) ...
 - level 4 -> UPDATE TABLE

Supported operations are getattr, readdir, read, open, mkdir, rmdir, create, write, unlink and truncate.

Level 1-3 entries are *always* directories, returning -ENOTDIR/-EPERM on level 4 actions.
Level 4 entries are *always* files, return -EISDIR/-EPERM on level 1 -- level 3 actions

Thoughts for consideration:
 - Read enhancement -> don't show NULL columns for particular record?
 - Support for more DB systems, like e.g. PostgreSQL etc.


NEWS
Fuse DB NEWS


src/fuse-db.c
/*
  MySQL FUSE Connector
  Designed and written by Michal Novotny <mignov@gmail.com> in 2010

  Based on:
  FUSE: Filesystem in Userspace
  Copyright (C) 2001-2007  Miklos Szeredi <miklos@szeredi.hu>

  This program can be distributed under the terms of the GNU GPL.
  See the file COPYING.
*/

#define DEBUG_MAIN

#ifdef DEBUG_MAIN
#define DPRINTF(fmt, ...) \
do { fprintf(stderr, "main: " fmt , ## __VA_ARGS__); } while (0)
#else
#define DPRINTF(fmt, ...) \
do {} while(0)
#endif

#include "fuse-db.h"

long flags = 0;

/* Connection parameters */
char *mServer   = NULL;
char *mUser     = NULL;
char *mPass     = NULL;
char *mMntPoint = NULL;
char *mLogFile  = NULL;
char *mPwdType  = "plain";

unsigned char *unbase64(char *input) {
    int size = 0;
    unsigned char *val = NULL;

    val = strdup( (char *)base64_decode(input, &size) );

    if (size > 3) {
        if (val[size-3] == '\r')
            val[size-3] = 0;
        else
        if (val[size-2] == '\n')
            val[size-2] = 0;
    }

    return val;
}

int countChars(const char *str, int c) {
    int num, i;

    num = 0;

    for (i = 0; i < strlen(str); i++)
        if (str[i] == c)
            num++;

    return num;
}

int getLevel(const char *path) {
    if (strcmp(path, "/") == 0)
        return 0;

    return countChars(path, '/');
}

int flagIsSet(int flag)
{
    return (flags & flag) ? 1 : 0;
}

int getErrorCode(int err, int code, int retTrue, int retFalse)
{
    if ((err == code) && flagIsSet(FLAG_CORRECT_CODES))
        return retTrue;
    
    return retFalse;
}

char *replace(char *input, char *what, char *with)
{
    char *out, *tmp, *str, *save, *token;
    int num = 0, size = 0;

    if ((what == NULL) || (strlen(what) == 0))
        return input;

    num = countChars((const char *)input, what[0]);
    size = (strlen(input) + num + 1);

    out = (char *)malloc( size * sizeof(char) );
    memset(out, 0, size);

    tmp = strdup( (char *)input );
    for (str = tmp; ; str = NULL) {
        token = strtok_r(str, what, &save);
        if (token == NULL)
            break;
        strcat(out, token);
        strcat(out, with);
    }
    out[ size - 1 ] = 0;

    return out;
}

char *escape(char *input)
{
    input = replace(input, "\'", "\\'");
    input = replace(input, "\\", "\\\\");

    return input;
}

char *getPathComponent(const char *path, int idx) {
    char *tmp, *str, *save, *token;
    int n = 0;

    tmp = strdup( (char *)path );
    for (str = tmp; ; str = NULL) {
        token = strtok_r(str, "/", &save);
        if (token == NULL)
            break;
        if (n == idx)
            return token;
        n++;
    }

    return NULL;
}

static struct fuse_operations fmysql_oper = {
    /* Directories/files listing */
    .getattr    = fmysql_getattr,
    .readdir    = fmysql_readdir,
    /* Read functions */
    .read       = fmysql_read,
    .open       = fmysql_open,
    /* Directory operations */
    .mkdir      = fmysql_mkdir,
    .rmdir      = fmysql_rmdir,
    /* Write operations */
    .create     = fmysql_create,
    .write      = fmysql_write,
    /* File operations */
    .unlink     = fmysql_rm,
    .truncate   = fmysql_truncate,
};

void dumpArgs() {
    if (!flagIsSet(FLAG_DEBUG))
        return;
    printf("\nDump argument settings:\n");
    printf("\tServer: %s\n", mServer);
    printf("\tUser: %s\n", mUser);
    if (flagIsSet(FLAG_DEBUGPWD))
        printf("\tPassword: %s\n", mPass);
    else
        printf("\tPassword: %s\n", mPass ? "Set" : "Not set");
    printf("\tPassword type: %s\n", mPwdType);
    printf("\tRead-only: %s\n", flagIsSet(FLAG_READONLY) ? "True" : "False");
    printf("\tLog file: %s\n", mLogFile);
    printf("\tMountpoint: %s\n", mMntPoint);
    printf("\tForce: %s\n", flagIsSet(FLAG_FORCE) ? "True" : "False");
    printf("\tUnmount: %s\n", flagIsSet(FLAG_UNMOUNT) ? "True" : "False");
    printf("\n");
}

void usage(char *name) {
    fprintf(stderr, "Syntax: %s --server <server> --user <user> --password <password> --password-type <type*1>\n"
                    "        --mountpoint <mountpoint> [--log-file <log-file>] [--debug] [--force-password-dump]\n"
                    "        [--force] [--use-correct-codes] [--read-only] [--unmount]\n\n"
                    "You can also use short version of the parameters by using the lowercase first letters except for\n"
                    "-t for password type and -g for debugging. Forcing the password dump will enforce dumping the\n"
                    "password in the debug output if enabled.\nFor the password-type you can use plain text type"
                    "which is the default or you can use 'b64' type\nthat specifies the password is in base64 encoded "
                    "format.\n", name);

    dumpArgs();
    exit(EXIT_FAILURE);
}

long parseArgs(int argc, char * const argv[]) {
    int option_index = 0, c;
    unsigned int retVal = 0;
    struct option long_options[] = {
        {"server", 1, 0, 's'},
        {"user", 1, 0, 'u'},
        {"password", 1, 0, 'p'},
        {"password-type", 1, 0, 't'},
        {"mountpoint", 1, 0, 'm'},
        {"log-file", 1, 0, 'l'},
        {"debug", 0, 0, 'g'},
        {"force-password-dump", 0, 0, 'd'},
        {"force", 0, 0, 'f'},
        {"unmount", 0, 0, 'n'},
        {"use-correct-codes", 0, 0, 'c'},
        {"read-only", 0, 0, 'r'},
        {0, 0, 0, 0}
    };

    char *optstring = "s:u:p:t:m:l:gfdn";

    while (1) {
        c = getopt_long(argc, argv, optstring,
                   long_options, &option_index);
        if (c == -1)
            break;

        switch (c) {
            case 's':
                mServer = strdup(optarg);
                break;
            case 'u':
                mUser = strdup(optarg);
                break;
            case 'p':
                mPass = strdup(optarg);
                break;
            case 'l':
                mLogFile = strdup(optarg);
                unlink(mLogFile);
                break;
            case 't':
                mPwdType = strdup(optarg);
                break;
            case 'm':
                mMntPoint = strdup(optarg);
                break;
            case 'g':
                retVal |= FLAG_DEBUG;
                break;
            case 'd':
                retVal |= FLAG_DEBUGPWD;
                break;
            case 'f':
                retVal |= FLAG_FORCE;
                break;
            case 'n':
                retVal |= FLAG_UNMOUNT;
                break;
            case 'c':
                retVal |= FLAG_CORRECT_CODES;
                break;
            case 'r':
                retVal |= FLAG_READONLY;
                break;
            default:
                usage(argv[0]);
        }
    }

    return retVal;
}

int unmount(char *binaryPath, char *mountpoint, int suppressMessages)
{
    char cmd[256], *tmp;
    FILE *fp;
    int pid, killed;

    tmp = (char *)malloc( 512 * sizeof(char) );
    memset(tmp, 0, 512);

    snprintf(cmd, sizeof(cmd), "umount %s 2> /dev/null", mountpoint);
    system(cmd);

    killed = 0;
    snprintf(cmd, sizeof(cmd), "ps -C %s -o pid=", binaryPath);
    fp = popen(cmd, "r");
    while (!feof(fp)) {
        fgets(tmp, 512, fp);
        if ((pid = atoi(tmp)) != getpid() ) {
            killed++;
            kill(pid, SIGKILL);
        }
    }
    fclose(fp);

    if (killed > 0) {
        if (!suppressMessages)
            printf("Path %s unmounted and process %d killed\n", mountpoint,
                    pid);

        return 0;
    }
    else
        if (!suppressMessages)
            printf("Cannot find any existing running instance of %s\n", binaryPath);

    return EXIT_FAILURE;
}

int main(int argc, char *argv[])
{
    int rc, i;
    struct statvfs vfs;

    if (getuid() != 0) {
       fprintf(stderr, "Error: You need to run %s as root!\n", argv[0]);
       return EXIT_FAILURE;
    }

    flags = parseArgs(argc, argv);

    if (!mServer || !mUser || !mPass || !mMntPoint)
        usage(argv[0]);

    if (strcmp(mPwdType, "b64") == 0)
        mPass = strdup(unbase64(mPass));

    dumpArgs();

    if (flagIsSet(FLAG_UNMOUNT) || (flagIsSet(FLAG_FORCE))) {
        rc = unmount( replace(argv[0], "./", ""), mMntPoint, flagIsSet(FLAG_FORCE) );

        if (flagIsSet(FLAG_UNMOUNT))
            return rc;

        if ((rc != 0) && (flagIsSet(FLAG_UNMOUNT)))
            fprintf(stderr, "Warning: Error unmounting mountpoint %s\n", mMntPoint);
    }

    if (statvfs(mMntPoint, &vfs) == 0) {
        if (vfs.f_fsid == 0) { /* Observed when mounted (FUSE mount) */
            fprintf(stderr, "Path %s seems to be already mounted. Cannot "
                            "continue. Quiting...\n", mMntPoint);
            return EXIT_FAILURE;
        }
    }

    if (mysql_init(&sql) == NULL)
        return EXIT_FAILURE;

    if (!mysql_real_connect(&sql, mServer, mUser, mPass, NULL, 0, NULL, 0)) {
        fprintf(stderr, "MySQL connection error: %s (%d)\n", mysql_error(&sql),
                mysql_errno(&sql));
        mysql_close(&sql);
        return EXIT_FAILURE;
    }

    /* Unset all the arguments for fuse_main */
    for (i = 1; i > argc; i++)
        free(argv[i]);
    argc = 2;
    /* Set only the mountpoint argument */
    argv[1] = mMntPoint;

    printf("Process %s started successfully\n", argv[0]);

    rc = fuse_main(argc, argv, &fmysql_oper, NULL);
    mysql_close(&sql);

    return rc;
}



src/fuse-db.h
#ifndef FUSE_DB_H
#define FUSE_DB_H

#define FUSE_USE_VERSION 26
#define EXT_LOG_SIZE     40960 /* 40 kiB */

#include <fuse.h>
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include <errno.h>
#include <stdarg.h>
#include <signal.h>
#include <sys/mount.h>
#include <mysql/mysql.h>

#define TYPE_NOENT      -1
#define TYPE_FILE       0
#define TYPE_DIR        1
#define TYPE_DIR_NOPK   2

#define FLAG_READONLY           4
#define FLAG_CORRECT_CODES      8
#define FLAG_UNMOUNT            16
#define FLAG_FORCE              32
#define FLAG_DEBUGPWD           64
#define FLAG_DEBUG              128

MYSQL sql;

/* Core functions */
unsigned char *unbase64(char *input);
int countChars(const char *str, int c);
int getLevel(const char *path);
int flagIsSet(int flag);
int getErrorCode(int err, int code, int retTrue, int retFalse);
char *replace(char *input, char *what, char *with);
char *escape(char *input);
char *getPathComponent(const char *path, int idx);

/* MySQL functions */
int getFieldNumber(MYSQL sql, char *qry, char *fieldName);
char *getValue(MYSQL sql, char *qry, char *fieldName, unsigned long long *numRows);
char *getPrimaryKeyName(MYSQL sql, char *table, int *error);
int getSize(MYSQL sql, char *path, int *error);
int getMySQLResults(MYSQL sql, char *qry, char *field, fuse_fill_dir_t filler, void *buf);
int isReadOnly(MYSQL sql, const char *path, char *tab);
int getType(char *path, int *error);
int fmysql_getattr(const char *path, struct stat *stbuf);
char *mysql_read(MYSQL sql, char *path, unsigned int *len);
int fmysql_read(const char *path, char *buf, size_t size, off_t offset, struct fuse_file_info *fi);
int fmysql_readdir(const char *path, void *buf, fuse_fill_dir_t filler, off_t offset, struct fuse_file_info *fi);
int fmysql_open(const char *path, struct fuse_file_info *fi);
int fmysql_mkdir(const char *path, mode_t mode);
int fmysql_rmdir(const char *path);
int fmysql_rm(const char *path);
int fmysql_create(const char *path, mode_t mode, struct fuse_file_info *fi);
int fmysql_truncate(const char *path, off_t size);
int fmysql_write(const char *path, const char *buf, size_t size, off_t offset, struct fuse_file_info *fi);

#endif


src/base64.c
/*  Copyright (c) 2006-2007, Philip Busch <broesel@studcs.uni-sb.de>
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions are met:
 *
 *   - Redistributions of source code must retain the above copyright notice,
 *     this list of conditions and the following disclaimer.
 *   - Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in the
 *     documentation and/or other materials provided with the distribution.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 *  POSSIBILITY OF SUCH DAMAGE.
 */

/**
 * @file
 * Base64 implementation.
 * @ingroup base64
 */

#include "fuse-db.h"

#define XX 100

/** @var base64_list
 *   A 64 character alphabet.
 *
 *   A 64-character subset of International Alphabet IA5, enabling
 *   6 bits to be represented per printable character.  (The proposed
 *   subset of characters is represented identically in IA5 and ASCII.)
 *   The character "=" signifies a special processing function used for
 *   padding within the printable encoding procedure.
 *
 *   \verbatim
    Value Encoding  Value Encoding  Value Encoding  Value Encoding
       0 A            17 R            34 i            51 z
       1 B            18 S            35 j            52 0
       2 C            19 T            36 k            53 1
       3 D            20 U            37 l            54 2
       4 E            21 V            38 m            55 3
       5 F            22 W            39 n            56 4
       6 G            23 X            40 o            57 5
       7 H            24 Y            41 p            58 6
       8 I            25 Z            42 q            59 7
       9 J            26 a            43 r            60 8
      10 K            27 b            44 s            61 9
      11 L            28 c            45 t            62 +
      12 M            29 d            46 u            63 /
      13 N            30 e            47 v
      14 O            31 f            48 w         (pad) =
      15 P            32 g            49 x
      16 Q            33 h            50 y
    \endverbatim
 */
static const char base64_list[] = \
        "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

static const int base64_index[256] = {
    XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX,
    XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX,
    XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,62, XX,XX,XX,63,
    52,53,54,55, 56,57,58,59, 60,61,XX,XX, XX,XX,XX,XX,
    XX, 0, 1, 2,  3, 4, 5, 6,  7, 8, 9,10, 11,12,13,14,
    15,16,17,18, 19,20,21,22, 23,24,25,XX, XX,XX,XX,XX,
    XX,26,27,28, 29,30,31,32, 33,34,35,36, 37,38,39,40,
    41,42,43,44, 45,46,47,48, 49,50,51,XX, XX,XX,XX,XX,
    XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX,
    XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX,
    XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX,
    XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX,
    XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX,
    XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX,
    XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX,
    XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX,
};

/** Encode a minimal memory block. This function encodes a minimal memory area
 *  of three bytes into a printable base64-format sequence of four bytes.
 *  It is mainly used in more convenient functions, see below.
 *
 * @attention This function can't check if there's enough space at the memory
 *            memory location pointed to by \c out, so be careful.
 *
 * @param out pointer to destination
 * @param in pointer to source
 * @param len input size in bytes (between 0 and 3)
 * @returns nothing
 *
 * @ingroup base64
 */
void base64_encode_block(unsigned char out[4], const unsigned char in[3], int len)
{
        out[0] = base64_list[ in[0] >> 2 ];
        out[1] = base64_list[ ((in[0] & 0x03) << 4) | ((in[1] & 0xf0) >> 4) ];
        out[2] = (unsigned char) (len > 1 ? base64_list[ ((in[1] & 0x0f) << 2) | ((in[2] & 0xc0) >> 6) ] : '=');
        out[3] = (unsigned char) (len > 2 ? base64_list[in[2] & 0x3f] : '=');
}

/** Decode a minimal memory block. This function decodes a minimal memory area
 *  of four bytes into its decoded equivalent. It is mainly used in more
 *  convenient functions, see below.
 *
 * @attention This function can't check if there's enough space at the memory
 *            memory location pointed to by \c out, so be careful.
 *
 * @param out pointer to destination
 * @param in pointer to source
 * @returns -1 on error (illegal character) or the number of bytes decoded
 *
 * @ingroup base64
 */
int base64_decode_block(unsigned char out[3], const unsigned char in[4])
{
        int i, numbytes = 3;
        char tmp[4];

        for(i = 3; i >= 0; i--) {
                if(in[i] == '=') {
                        tmp[i] = 0;
                        numbytes = i - 1;
                } else {
                        tmp[i] = base64_index[ (unsigned char)in[i] ];
                }
                
                if(tmp[i] == XX)
                        return(-1);
        }

        out[0] = (unsigned char) (  tmp[0] << 2 | tmp[1] >> 4);
        out[1] = (unsigned char) (  tmp[1] << 4 | tmp[2] >> 2);
        out[2] = (unsigned char) (((tmp[2] << 6) & 0xc0) | tmp[3]);

        return(numbytes);
}

/** Compute size of needed storage for encoding. This function computes the
 *  \e exact size of a memory area needed to hold the result of an encoding
 *  operation, not including the terminating null character.
 *
 * @param len input size
 * @returns output size
 *
 * @ingroup base64
 */
size_t base64_encoded_size(size_t len)
{
        return(((len + 2) / 3) * 4);
}

/** Compute size of needed storage for decoding. This function computes the
 *  \e estimated size of a memory area needed to hold the result of a decoding
 *  operation, not including the terminating null character. Note that this
 *  function may return up to two bytes more due to the nature of Base64.
 *
 * @param len input size
 * @returns output size
 *
 * @ingroup base64
 */
size_t base64_decoded_size(size_t len)
{
        return((len / 4) * 3);
}

/** Encode an arbitrary size memory area. This function encodes the first
 *  \c len bytes of the contents of the memory area pointed to by \c in and
 *  stores the result in the memory area pointed to by \c out. The result will
 *  be null-terminated.
 *
 * @attention This function can't check if there's enough space at the memory
 *            memory location pointed to by \c out, so be careful.
 *
 * @param out pointer to destination
 * @param in pointer to source
 * @param len input size in bytes
 * @returns nothing
 *
 * @ingroup base64
 */
void base64_encode_binary(char *out, const unsigned char *in, size_t len)
{
        int size;
        size_t i = 0;
        
        while(i < len) {
                size = (len-i < 4) ? len-i : 4;
                
                base64_encode_block((unsigned char *)out, in, size);

                out += 4;
                in  += 3;
                i   += 3;
        }

        *out = '\0';
}

/** Decode an arbitrary size memory area. This function decodes the
 *  base64-string pointed to by \c in and stores the result in the memory area
 *  pointed to by \c out. The result will \e not be null-terminated.
 *
 * @attention This function can't check if there's enough space at the memory
 *            memory location pointed to by \c out, so be careful.
 *
 * @param out pointer to destination
 * @param in pointer to source
 * @returns -1 on error (illegal character) or the number of bytes decoded
 *
 * @ingroup base64
 */
int base64_decode_binary(unsigned char *out, const char *in)
{
        size_t len = strlen(in), i = 0;
        int numbytes = 0;

        while(i < len) {
                if((numbytes += base64_decode_block(out, (unsigned char *)in)) < 0)
                        return(-1);

                out += 3;
                in  += 4;
                i   += 4;
        }

        return(numbytes);
}

/** Encode a string. This is a convenience function. It encodes the first
 *  \c size bytes of the string pointed to by \c in, stores the null-terminated
 *  result in a newly created memory area and returns a pointer to it.
 *
 * @attention After a call to base64_encode(), you have to free() the result
 *  yourself.
 *
 * @param in pointer to string
 * @param size strlen
 * @returns NULL on error (not enough memory) or a pointer to the encoded result
 *
 * @ingroup base64
 */
unsigned char *base64_encode(const char *in, size_t *size)
{
        unsigned char *out;
        size_t outlen;
	size_t esize;

        if((in == NULL) || (size == NULL))
                return(NULL);

        esize = *size;
        if(esize == 0)
                esize = strlen(in);

        outlen = base64_encoded_size(esize);

        if((out = (unsigned char *)malloc(sizeof(unsigned char) * (outlen + 1))) == NULL)
                return(NULL);

        base64_encode_binary(out, (unsigned char *)in, esize);

	*size = outlen;

        return(out);
}

/** Decode a string. This is a convenience function. It decodes the
 *  null-terminated string pointed to by \c in, stores the result in a newly
 *  created memory area and returns a pointer to it. The result will be
 *  null-terminated.
 *
 * @attention After a call to base64_decode(), you have to free() the result
 *  yourself.
 *
 * @param in pointer to string
 * @returns NULL on error (not enough memory) or a pointer to the decoded result
 *
 * @ingroup base64
 */
unsigned char *base64_decode(const char *in, size_t *size)
{
        unsigned char *out;
        size_t outlen;
        int numbytes;
        
	if (size == NULL)
		return NULL;

	if (*size == 0)
		*size = strlen(in);

        outlen = base64_decoded_size(*size);

        if((out = (unsigned char *)malloc(sizeof(unsigned char) * (outlen + 1))) == NULL)
                return(NULL);

        if((numbytes = base64_decode_binary((unsigned char *)out, in)) < 0) {
                free(out);
                return(NULL);
        }
        
        *(out + numbytes) = '\0';
	*size = outlen;
        
        return(out);
}

